### DISCUSSION TITLE
Transformers using device_map="auto" in fastapi with workers>=2 will report an error with multiple graphics cards

### DISCUSSION THREAD

**Discussion by ciaoyizhen** [fastapi/fastapi#11862] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
import torch
from torch import nn
from fastapi import FastAPI
from transformers import AutoModel, AutoTokenizer

app = FastAPI()
tokenizer = AutoTokenizer.from_pretrained('maidalun1020/bce-embedding-base_v1')
model = AutoModel.from_pretrained('maidalun1020/bce-embedding-base_v1', device_map="auto")
print("success")
@app.get("/")
def index():
    return {"message": "hello"}

if __name__ == "__main__":
    uvicorn.run("main:app", workers=2, host="0.0.0.0", port=54441)
```

### Description

When I start this code with cuda, the child process died and then restarted and died again.
On a cuda as well as a cpu as well as mps it is all running.
But two cuda is error.

### Operating System

Linux

### Operating System Details

V100 * 4
Centos7

### FastAPI Version

0.111.0

### Pydantic Version

2.7.3

### Python Version

3.10.12

**Comment by YuriiMotov** [comment]
Do you see any error messages in the console or logs?

**Reply by ciaoyizhen** [reply]
INFO - uvicorn running on http://0.0.0.0:54441 (Press CTRL+C to quit)
INFO - Started parent process [59295]
INFO - Waiting for child process [67752]
INFO - Child process [67752] died
INFO - Waiting for child process [67753]
INFO - Child process [67753] died


LIke this.
Keep repeating these started and died.

**Reply by ciaoyizhen** [reply]
If you have multiple gpus try running my code above, if there is only a single gpu or only cpu or on mps then the code above is fine.

**Reply by YuriiMotov** [reply]
First we need to make sure it's a FastAPI or Uvicorn-related issue.
Have you tried to load two instances of that model without FastAPI and Uvicorn?
Something like:
```
from transformers import AutoModel, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('maidalun1020/bce-embedding-base_v1')
model = AutoModel.from_pretrained('maidalun1020/bce-embedding-base_v1', device_map="auto")

tokenizer_2 = AutoTokenizer.from_pretrained('maidalun1020/bce-embedding-base_v1')
model_2 = AutoModel.from_pretrained('maidalun1020/bce-embedding-base_v1', device_map="auto")

print("success")
```
I don't have any experience working with `transformers`. May be you should load this models in different processes.

**Reply by ciaoyizhen** [reply]
transformers is an NLP package that allows you to quickly turn on an NLP model.
Here is a BCEmbedding model running.
I've tried it without device_map (which is in fact an implementation of auto-detection of the model storage location on multiple cards), i.e. with a single card, and everything works fine!

**Reply by YuriiMotov** [reply]
Have you tried to load two instances of that model without FastAPI and Uvicorn?
Just create two processes with multiprocessing or run in a single process

**Reply by ciaoyizhen** [reply]
```python
import time
from transformers import AutoModel, AutoTokenizer
from concurrent.futures import ProcessPoolExecutor, as_completed

def target():
    tokenizer = AutoTokenizer.from_pretrained('maidalun1020/bce-embedding-base_v1')
    model = AutoModel.from_pretrained('maidalun1020/bce-embedding-base_v1', device_map="auto")
    return model, tokenizer

models = []
tokenizers = []
with ProcessPoolExecutor(max_workers=2) as exe:
    futures = []
    for i in range(2):
        futures.append(exe.submit(target))
        
    for future in as_completed(futures):
        model, tokenizer = future.result()
        models.append(model)
        tokenizers.append(tokenizer)
        print("success")
        
time.sleep(3600)
```

this code is OK.

**Reply by YuriiMotov** [reply, marked as answer]
Try running this:
```
from contextlib import asynccontextmanager
import time
import uvicorn
import torch
from torch import nn
from fastapi import FastAPI
from transformers import AutoModel, AutoTokenizer
from fastapi.concurrency import run_in_threadpool

def load_models():
    tokenizer = AutoTokenizer.from_pretrained('maidalun1020/bce-embedding-base_v1')
    model = AutoModel.from_pretrained('maidalun1020/bce-embedding-base_v1', device_map="auto")
    return tokenizer, model

@asynccontextmanager
async def lifespan(app: FastAPI):
    try:
        tokenizer, model = await run_in_threadpool(load_models)
    except Exception as e:
        print(e)
        raise
    print("Model loaded")
    yield {"model": model, "tokenizer": tokenizer}

app = FastAPI(lifespan=lifespan)

@app.get("/")
def index():
    return {"message": "hello"}

if __name__ == "__main__":
    uvicorn.run("t:app", workers=2, host="0.0.0.0", port=54441, log_level="info")
```

**Reply by ciaoyizhen** [reply]
wow!!!!  It's OK.  Thank you very much !!!!!!!!
